### PR TITLE
Ers quick choice enhanced

### PR DESCRIPTION
**Pull request by ericrsmith35**
12/30/19 -  Eric Smith -    Version 1.14
                            Added visual textbox selection with or without icons
                            Fixed error handling when Required is set to True

### CODE REVIEW

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js [hunk 1]
```diff
@@ -1,167 +1,215 @@
...
+					value: key.value
+				});
+			});
+
+			this.picklistOptionsStorage = picklistOptions;
+			console.log("displayMode is" + this.displayMode);
+
+			// Visual Picker Selection
+			if (this.displayMode === "Visual") {
+				this.showVisual = true;
```
Comment (alexed1) [inline flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: This boolean approach isn't really scalable and wasn't a good initial design decision. can we just remove it? the existing displayMode should be sufficient, I suspect.

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml [hunk 2]
```diff
@@ -11,13 +11,16 @@
             <property name="value" label="Value" type="String"></property>
             <property name="choiceValues" label="Choice Values (String Collection)" type="String[]"></property>
             <property name="choiceLabels" label="Choice Labels (String Collection)" type="String[]"></property>
+            <property name="choiceIcons" label="Choice Icons (String Collection)" type="String[]"></property>
+            <property name="includeIcons" label="Include Icons in Text Box?" type="Boolean"></property>
```
Comment (alexed1) [inline flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml]: always look for the most evocative language when naming variables. In this case, showing vs not showing is clearer than including vs not including, no?

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js [hunk 3]
```diff
@@ -1,167 +1,215 @@
...
+						index +=1;
+					});
+					break;
+
+				default:
+			}
+			this.options = options;
+			this.items = items;
+		} else {
+			console.log(
```
Comment (alexed1) [inline flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: keep console.log lines to one line where possible.

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js [hunk 4]
```diff
@@ -1,167 +1,215 @@
...
+				"SmartChoiceFSC: Need a valid Input Mode value. Didn't get one"
+			);
+			throw new Error(
```
Comment (alexed1) [inline flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: likewise, error message throws should generally be kept streamlined and one line.

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js [hunk 5]
```diff
@@ -1,167 +1,215 @@
...
+			return {
+				isValid: false,
+				errorMessage: errorMessage
+			};
+		}
+
+		return { isValid: true };
+	}
+
+	handleChange(event) {
```
Comment (alexed1) [inline flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: notice how handleChange and handleRadio are nearly duplicates? pro developers develop a spidey sense for things like that and treat them as opportunities to combine functionality to DRY things up. in this case, can you combine them into a single handleChange that sets the selectedValue appropriately based on something in the event? (or is event.detail.value definitely different from event.target value? )

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js (hunk 1 above)
Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: showVisual and showRadio are conditional properties used in the template.  They are set based on the displayMode attibute.  I don't think a boolean expression is valid in the html template.

I did, however, remove the redundant setting of these in getPicklistValues and the uneccessary clearing of showRadio when displayMode is "Visual".

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml (hunk 2 above)
Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml]: I've changed the variable to showIcons and updated the documentation

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js (hunk 3 above)
Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: It looks like the auto formatting in Vscode split those into multiple lines.  I've changed them back.

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js (hunk 4 above)
Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: Same Vscode issue on formatting

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js (hunk 5 above)
Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: They are different for input type=radio vs the lightning-combobox or radio-group.  I removed handleRadio and changed the assignment in handleChange to this:
this.selectedValue = (this.showVisual) ? event.target.value : event.detail.value;

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml (hunk 2 above)
Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml]: because of this change and the managed package i'm unable to deploy these changes to hostquick-dev

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js (hunk 1 above)
Comment (alexed1) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: I see what you mean.

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js (hunk 4 above)
Comment (alexed1) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: fwiw my vscode doesn't do that expansion so it's something that's setting-specific..... not a big deal....

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js (hunk 5 above)
Comment (alexed1) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js]: perfect resolution. cleaner, shorter code.

File: flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml (hunk 2 above)
Comment (alexed1) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml]: if we revert this change, can you then deploy to that org and update the existing managed package?

Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml]: yes

Comment (alexed1) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml]: ok go for it

Comment (ericrsmith35) [reply flow_screen_components/quickChoiceFSCproject/force-app/main/default/lwc/quickChoiceFSC/quickChoiceFSC.js-meta.xml]: See new pull request #237

**Comment by shaktigopal** [comment]
@alexed1 @ericrsmith35 
Hey Eric,
I am using quickchoiceFSC (_https://github.com/alexed1/LightningFlowComponents/tree/master/flow_screen_components/quickChoiceCPE/force-app/main/default/lwc/quickChoiceFSC_) in a flow.

I am using this component twice, on same screen, tagged with two separate picklist fields. If in the first component i give API name "cmpProduct". And I use "cmpProduct.value" in the visible component visibility of second component (e.g. show second component if "cmpProduct.value" = "Yes"), then it works fine in the normal flow process. BUT if without selecting anything in the first picklist, I click on Next and then click on Previous to select the value Yes on first component, then it gives the error.
Uncaught Error during LWC component connect phase: [SmartChoiceFSC: Need a valid Input Mode value. Didn't get one].

Basically "SmartChoiceFSC: Need a valid Input Mode value. Didn't get one" is being thrown from connectedCallback() method. It is thrown if input variables have not got the value. So yes all the input property variables for the second component comes blank if it is tried to make visible on screen after coming back on the screen by clicking Next & Previous.

Any help would be appreciable.

Thanks
Shakti

**Comment by ericrsmith35** [comment]
Shakti,

This post by Alex will better explain some of the behaviors when moving
back and forth between Flow screens.

https://unofficialsf.com/optimizing-lightning-component-inputs-and-outputs-in-flow-screens/


Eric
